python/semantic_kernel/planning/stepwise_planner/stepwise_planner.py

```diff
@@ -0,0 +1,459 @@
+
+        return context
+
+    def parse_result(self, input: str):
+        result = SystemStep(original_response=input)
+
+        # Extract final answer
+        final_answer_match = re.search(S_FINAL_ANSWER_REGEX, input)
+
+        if final_answer_match:
+            result.final_answer = final_answer_match.group(1).strip()
+            return result
+
+        # Extract thought
+        thought_match = re.search(S_THOUGHT_REGEX, input)
+
+        if thought_match:
+            result.thought = thought_match.group(0).strip()
+        elif not input.contains(ACTION):
+            result.thought = input
+        else:
+            raise ValueError("Unexpected input format")
+
+        result.thought = result.thought.replace(THOUGHT, "").strip()
+
+        # Extract action
+        action_match = re.search(S_ACTION_REGEX, input)
+
+        if action_match:
+            action_json = action_match.group(1).strip()
+
+            try:
+                system_step_results = json.loads(action_json)
+
+                if system_step_results is None or len(system_step_results) == 0:
+                    result.observation = (
+                        f"System step parsing error, empty JSON: {action_json}"
+                    )
+                else:
+                    result.action = system_step_results["action"]
+                    result.action_variables = system_step_results["action_variables"]
+            except Exception:
+                result.observation = (
+                    f"System step parsing error, invalid JSON: {action_json}"
+                )
+
+        if is_null_or_empty(result.thought) and is_null_or_empty(result.action):
+            result.observation = (
+                "System step error, no thought or action found.",
+                "Please give a valid thought and/or action.",
+            )
+
+        return result
+
+    def add_execution_stats_to_context(
+        self, steps_taken: List[SystemStep], context: SKContext
+    ):
+        context.variables.set("step_count", str(len(steps_taken)))
+        context.variables.set(
+            "steps_taken", json.dumps([s.__dict__ for s in steps_taken], indent=4)
+        )
+
+        action_counts: Dict[str, int] = {}
+        for step in steps_taken:
+            if is_null_or_empty(step.action):
+                continue
+
+            current_count = action_counts.get(step.action, 0)
+            action_counts[step.action] = current_count + 1
+
+        skill_call_list_with_counts = [
+            f"{skill}({action_counts[skill]})" for skill in action_counts
+        ]
+        skill_call_list_with_counts = ", ".join(skill_call_list_with_counts)
+        skill_call_count_str = str(sum(action_counts.values()))
+
+        context.variables.set(
+            "skill_count", f"{skill_call_count_str} ({skill_call_list_with_counts})"
+        )
+
+    def create_scratch_pad(self, question: str, steps_taken: List[SystemStep]) -> str:
```

Comment:
Just a note: in dotnet I'm looking to move away from this logic and manage the message stack in a ChatHistory object which can then be used with both text completion and chat completion models. This is fine for now though. THanks!

Reply:
@lemillermicrosoft 
I think it's a good change, and it would be nice to see synchronization from other PRs when that part is updated.